## What Is Market Basket Analysis?

* Market Basket Analysis is a technique used by retailers to discover relationships between products that customers buy. For example:

        “If a customer buys bread and butter, they are likely to buy milk.”

## What Is Association Rule Mining?

* Market Basket Analysis is one of the key techniques used by large retailers to uncover associations between items. It works by looking for combinations of items that occur together frequently in transactions. To put it another way, it allows retailers to identify relationships between the items that people buy.

* Association Rules are widely used to analyze retail basket or transaction data and are intended to identify strong rules discovered in transaction data using measures of interestingness, based on the concept of strong rules.
* Association rule mining finds interesting associations (rules) among items in large datasets. These rules are of the form:

    * IF {item1, item2} THEN {item3}
    
    * Example Rule:
    
        {Bread, Butter} → {Milk}
    
    * This means that customers who buy bread and butter are also likely to buy milk.

## Key Metrics:

* Support:

    * How frequently an itemset appears in the dataset. This says how popular an itemset is, as measured by the proportion of transactions in which an itemset appears.
    
        Support(A) = (Transactions containing A) / (Total transactions)

* Confidence:
    
    * This says how likely item Y is purchased when item X is purchased, expressed as {X -> Y}. This is measured by the proportion of transactions with item X, in which item Y also appears.
    
        Confidence(A → B) = Support(A ∪ B) / Support(A)

* Lift:

    * Measures how much more likely B is to happen given A.
    
        Lift(A → B) = Confidence(A → B) / Support(B)
    
    * If Lift > 1, the rule is strong.

## What Is Apriori Algorithm?

* Apriori is an algorithm for frequent itemset mining and association rule learning over relational databases. It proceeds by identifying the frequent individual items in the database and extending them to larger and larger item sets as long as those item sets appear sufficiently often in the database. The frequent itemsets determined by Apriori can be used to determine association rules which highlight general trends in the database: this has applications in domains such as market basket analysis.

#### An example of Association Rules

* Assume there are 100 customers
    * 10 of them bought milk, 8 bought butter and 6 bought both of them.
    * bought milk => bought butter
        * support = P(Milk & Butter) = 6/100 = 0.06
        * confidence = support/P(Butter) = 0.06/0.08 = 0.75
        * lift = confidence/P(Milk) = 0.75/0.10 = 7.5

* Note: this example is extremely small. In practice, a rule needs the support of several hundred transactions, before it can be considered statistically significant, and datasets often contain thousands or millions of transactions.

### PRACTICAL CODING EXAMPLE

In [15]:
# pip install mlxtend pandas

In [3]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder

In [5]:
# Sample data - list of transactions
dataset = [
    ['Milk', 'Bread', 'Butter'],
    ['Milk', 'Bread'],
    ['Bread', 'Butter'],
    ['Milk', 'Bread', 'Butter', 'Eggs'],
    ['Milk', 'Eggs'],
    ['Bread', 'Eggs']
]

In [7]:
# Convert to one-hot encoded DataFrame
te = TransactionEncoder()
te_data = te.fit(dataset).transform(dataset)
df = pd.DataFrame(te_data, columns=te.columns_)

print("One-hot encoded transaction data:\n")
df

One-hot encoded transaction data:



Bread  Butter   Eggs   Milk
0   True    True  False   True
1   True   False  False   True
2   True    True  False  False
3   True    True   True   True
4  False   False   True   True
5   True   False   True  False

In [9]:
from mlxtend.frequent_patterns import apriori

# Get frequent itemsets with minimum support of 0.5 (50%)
frequent_itemsets = apriori(df, min_support=0.5, use_colnames=True)

print("\nFrequent Itemsets:\n")
print(frequent_itemsets)


Frequent Itemsets:

    support         itemsets
0  0.833333          (Bread)
1  0.500000         (Butter)
2  0.500000           (Eggs)
3  0.666667           (Milk)
4  0.500000  (Bread, Butter)
5  0.500000    (Bread, Milk)


In [11]:
from mlxtend.frequent_patterns import association_rules

# Generate rules with min confidence of 0.6
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.6)

# Show selected columns
print("\nAssociation Rules:\n")
print(rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']])



Association Rules:

  antecedents consequents  support  confidence  lift
0     (Bread)    (Butter)      0.5        0.60   1.2
1    (Butter)     (Bread)      0.5        1.00   1.2
2     (Bread)      (Milk)      0.5        0.60   0.9
3      (Milk)     (Bread)      0.5        0.75   0.9


antecedents: The "IF" part of the rule (items already bought)

consequents: The "THEN" part of the rule (item likely to be bought next)

support: Fraction of total transactions that contain both antecedent and consequent

confidence: Probability of buying consequent given the antecedent

lift: Strength of the rule compared to random chance (Lift > 1 is good)

### Row-by-Row Interpretation:
🔸 Row 0: (Bread) → (Butter)

    Support = 0.50 → 50% of all transactions contain both Bread and Butter.
    
    Confidence = 0.60 → Among transactions with Bread, 60% also include Butter.
    
    Lift = 1.2 → Buying Bread increases the likelihood of buying Butter by 20% compared to random chance.
    
    Interpretation: Customers who buy Bread have a moderate tendency to also buy Butter.

🔸 Row 1: (Butter) → (Bread)

    Support = 0.50 → Same 50% of transactions contain both.
    
    Confidence = 1.00 → 100% of transactions with Butter also have Bread.
    
    Lift = 1.2 → Strong positive association.
    
    Interpretation: Every customer who buys Butter also buys Bread. Strong rule.

🔸 Row 2: (Bread) → (Milk)

    Support = 0.50 → 50% of transactions have both Bread and Milk.
    
    Confidence = 0.60 → 60% of people who buy Bread also buy Milk.
    
    Lift = 0.9 → Less than 1, so slightly weaker than random chance.
    
    Interpretation: Buying Bread doesn’t strongly predict buying Milk. The rule is not very useful.

🔸 Row 3: (Milk) → (Bread)

    Support = 0.50 → 50% have both Milk and Bread.
    
    Confidence = 0.75 → 75% of people who buy Milk also buy Bread.
    
    Lift = 0.9 → Again, weaker than random.
    
    Interpretation: While a good portion of Milk-buyers also buy Bread, it's not significantly better than chance.